libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { Assertion, assert } from 'botbuilder-stdlib';
+
+import {
+    ExtendedUserTokenProvider,
+    Middleware,
+    Storage,
+    StoreItem,
+    TurnContext,
+    tokenExchangeOperationName,
+} from 'botbuilder-core';
+
+import {
+    ActivityTypes,
+    Channels,
+    StatusCodes,
+    TokenExchangeInvokeRequest,
+    TokenExchangeInvokeResponse,
+    TokenResponse,
+} from 'botframework-schema';
+
+function getStorageKey(context: TurnContext): string {
+    const activity = context.activity;
+
+    const channelId = activity.channelId;
+    if (!channelId) {
+        throw new Error('invalid activity. Missing channelId');
+    }
+
+    const conversationId = activity.conversation?.id;
+    if (!conversationId) {
+        throw new Error('invalid activity. Missing conversation.id');
+    }
+
+    const value = activity.value;
+    if (!value?.id) {
+        throw new Error('Invalid signin/tokenExchange. Missing activity.value.id.');
+    }
+
+    return `${channelId}/${conversationId}/${value.id}`;
+}
+
+async function sendInvokeResponse(context: TurnContext, body: unknown = null, status = StatusCodes.OK): Promise<void> {
+    await context.sendActivity({
+        type: ActivityTypes.InvokeResponse,
+        value: { body, status },
+    });
+}
+
+const assertExchangeToken: Assertion<Pick<ExtendedUserTokenProvider, 'exchangeToken'>> = (val, path) => {
+    assert.unsafe.castObjectAs<ExtendedUserTokenProvider>(val, path);
+    assert.func(val.exchangeToken, path.concat('exchangeToken'));
+};
```

Comment:
Thanks! I know we could use this assertion in the OAuthPrompt and probably OAuthInput.